navcoin/navcoin-core
# [TEST] 639 CFund Tests

Related to Issue #639 

Adding the foundations for a comprehensive test suite for the NavCoin DAO / CFund.

I've removed the "Work in Progress" label on this PR. Instead of waiting for every test to be written i think its better to merge what is currently there and then slowly add tests otherwise we could be waiting forever until every possible scenario is covered.

It's important to understand that what is in this PR is just some simple building blocks, once the basics are covered we can use these building blocks to construct ever increasingly complex test scenarios and perhaps eventually even automate/randomise test scenarios to see if anything breaks after sequences of events which we might not think to test.

I've merged in the latest master and all the tests i've written so far are passing. Can I please get some reviews and comments or approvals.

The current test scenarios that are covered can be found in the [test plan](https://github.com/navcoin/navcoin-core/blob/87a744a1a322331fb5b9fe10d01f96eda5edd9fc/qa/rpc-tests/dao/test-plan.md) but are summarised below;

## Scenarios

### Scenario 001

| ID          |  CFUND_003 |
| ----------- | -----: |
| Reporter    | @proletesseract |
| Author      | @proletesseract |
| Satus       | Done |
| Description | It should create a proposal and let it expire |
| File        | `./001-proposal-expired.py`

1. **Given** I have created a proposal
2. **When** The proposal is not voted on
3. **Then** the proposal should expire 

#### Test Steps
- Activate CFund
- Donate to the CFund
- Create a proposal
- Do not cast any votes
- End the full round of voting cycles
- Check proposal expired

### Scenario 002

| ID          |  CFUND_002 |
| ----------- | -----: |
| Reporter    | @proletesseract |
| Author      | @proletesseract |
| Satus       | Done |
| Description | It should create a proposal and the network should reject the proposal |
| File        | `./002-proposal-rejected.py`

1. **Given** I have created a proposal
2. **When** The proposal receives the consensus of `no` votes
3. **Then** the proposal should be rejected 

#### Test Steps
- Activate CFund
- Donate to the CFund
- Create a proposal
- Votes `no` on the proposal
- End the full round of voting cycles
- Check proposal is rejected

### Scenario 003

| ID          |  CFUND_003 |
| ----------- | -----: |
| Reporter    | @proletesseract |
| Author      | @proletesseract |
| Satus       | Done |
| Description | It should create a proposal and the network should accept the proposal |
| File        | `./003-proposal-accepted.py`

1. **Given** I have created a proposal
2. **When** The proposal receives the consensus of `yes` votes
3. **Then** the proposal should be accepted 

#### Test Steps
- Activate CFund
- Donate to the CFund
- Create a proposal
- Votes `yes` on the proposal
- End the full round of voting cycles
- Check proposal is rejected

### Scenario 004

| ID          |  CFUND_004 |
| ----------- | -----: |
| Reporter    | @proletesseract |
| Author      | @proletesseract |
| Satus       | Done |
| Description | It should fail to create a payment request for an expired proposal |
| File        | `./004-proposal-expired-preq.py`

1. **Given** I have an expired proposal
2. **When** A payment request is made
3. **Then** the payment request should not exist

#### Test Steps
- Activate CFund
- Donate to the CFund
- Create a proposal
- Do not cast any votes
- End the full round of voting cycles
- Submit a payment request
- Check payment request does not exist

### Scenario 005

| ID          |  CFUND_005 |
| ----------- | -----: |
| Reporter    | @proletesseract |
| Author      | @proletesseract |
| Satus       | Done |
| Description | It should fail to create a payment request for a rejected proposal |
| File        | `./005-proposal-expired-preq.py`

1. **Given** I have a rejected proposal
2. **When** A payment request is made
3. **Then** the payment request should not exist

#### Test Steps
- Activate CFund
- Donate to the CFund
- Create a proposal
- Votes `no` on the proposal
- End the full round of voting cycles
- Submit a payment request
- Check payment request does not exist

### Scenario 006

| ID          |  CFUND_006 |
| ----------- | -----: |
| Reporter    | @proletesseract |
| Author      | @proletesseract |
| Satus       | Done |
| Description | It should create a payment request for an accepted proposal |
| File        | `./006-proposal-accepted-preq.py`

1. **Given** I have an accepted proposal
2. **When** A payment request is made
3. **Then** the payment request should exist

#### Test Steps
- Activate CFund
- Donate to the CFund
- Create a proposal
- Votes `yes` on the proposal
- End the full round of voting cycles
- Submit a payment request
- Check payment request exists
